src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfo_Metadata.cs

```diff
@@ -66,73 +91,98 @@ private static string GetMetadataNameWithoutBackticks(MetadataReader reader, Str
             }
         }
 
-        public static ValueTask<SymbolTreeInfo> GetInfoForMetadataReferenceAsync(
-            Solution solution, PortableExecutableReference reference, CancellationToken cancellationToken)
-        {
-            var checksum = GetMetadataChecksum(solution.Services, reference, cancellationToken);
-            return GetInfoForMetadataReferenceAsync(solution, reference, checksum, cancellationToken);
-        }
-
         /// <summary>
         /// Produces a <see cref="SymbolTreeInfo"/> for a given <see cref="PortableExecutableReference"/>.
         /// Note:  will never return null;
         /// </summary>
+        /// <param name="checksum">Optional checksum for the <paramref name="reference"/> (produced by <see
+        /// cref="GetMetadataChecksum"/>).  Can be provided if already computed.  If not provided it will be computed
+        /// and used for the <see cref="SymbolTreeInfo"/>.</param>
         [PerformanceSensitive("https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1224834", OftenCompletesSynchronously = true)]
-        private static async ValueTask<SymbolTreeInfo> GetInfoForMetadataReferenceAsync(
+        public static async ValueTask<SymbolTreeInfo> GetInfoForMetadataReferenceAsync(
             Solution solution,
             PortableExecutableReference reference,
-            Checksum checksum,
+            Checksum? checksum,
             CancellationToken cancellationToken)
         {
-            var metadataId = GetMetadataIdNoThrow(reference);
-            if (metadataId == null)
-                return CreateEmpty(checksum);
+            checksum ??= GetMetadataChecksum(solution.Services, reference, cancellationToken);
 
-            if (s_metadataIdToInfo.TryGetValue(metadataId, out var infoTask))
+            if (s_peReferenceToInfo.TryGetValue(reference, out var infoTask))
             {
                 var info = await infoTask.GetValueAsync(cancellationToken).ConfigureAwait(false);
-                if (info.Checksum == checksum)
-                    return info;
+                Contract.ThrowIfTrue(info.Checksum != checksum, "How could the info stored for a particular PEReference now have a different checksum?");
+                return info;
```

Comment:
this contract needs to be true.  if we're keying by PERef we should always get the same checksum (as that is determined purely by the PERef)